File: app/providers/claude.py
import json
import time
import uuid
from typing import AsyncIterator

import httpx

from .. import store
from ..harvester import cookie_header

NAME = "claude"
MODELS = [
    "claude-opus-4-5",
    "claude-sonnet-4-5",
    "claude-haiku-4-5",
    "claude-3-5-sonnet-20241022",
    "claude-3-5-haiku-20241022",
]

_BASE = "https://claude.ai"
_DOMAINS = ("claude.ai",)


def _headers(session: dict) -> dict:
    return {
        "User-Agent": session.get("user_agent") or "Mozilla/5.0",
        "Accept": "text/event-stream, application/json",
        "Accept-Language": "en-US,en;q=0.9",
        "Content-Type": "application/json",
        "Origin": _BASE,
        "Referer": f"{_BASE}/chats",
        "Cookie": cookie_header(session, _DOMAINS),
    }


async def _org_id(client: httpx.AsyncClient, session: dict) -> str:
    r = await client.get(f"{_BASE}/api/organizations", headers=_headers(session))
    r.raise_for_status()
    orgs = r.json()
    if not orgs:
        raise RuntimeError("no claude organizations found on this account")
    for o in orgs:
        caps = o.get("capabilities") or []
        if "chat" in caps:
            return o["uuid"]
    return orgs[0]["uuid"]


def _flatten(messages: list[dict]) -> tuple[str, str]:
    sys_parts = [m["content"] for m in messages if m.get("role") == "system" and isinstance(m.get("content"), str)]
    convo = []
    for m in messages:
        role = m.get("role")
        if role == "system":
            continue
        content = m.get("content")
        if isinstance(content, list):
            content = "".join(p.get("text", "") for p in content if isinstance(p, dict) and p.get("type") == "text")
        if not isinstance(content, str):
            content = str(content)
        tag = {"user": "Human", "assistant": "Assistant"}.get(role, role.title())
        convo.append(f"{tag}: {content}")
    prompt = "\n\n".join(convo)
    if not prompt.endswith("Assistant:"):
        prompt += "\n\nAssistant:"
    return ("\n\n".join(sys_parts), prompt)


async def stream(req: dict) -> AsyncIterator[dict]:
    session = store.load_session("claude")
    if not session:
        raise RuntimeError("claude not connected — open the dashboard and click Connect Claude")

    model = req.get("model", "claude-sonnet-4-5")
    system, prompt = _flatten(req.get("messages", []))

    async with httpx.AsyncClient(http2=True, timeout=httpx.Timeout(120.0, read=300.0)) as client:
        org = await _org_id(client, session)

        conv_id = str(uuid.uuid4())
        r = await client.post(
            f"{_BASE}/api/organizations/{org}/chat_conversations",
            headers=_headers(session),
            json={"uuid": conv_id, "name": ""},
        )
        r.raise_for_status()

        payload = {
            "prompt": prompt,
            "parent_message_uuid": "00000000-0000-4000-8000-000000000000",
            "timezone": "America/Los_Angeles",
            "attachments": [],
            "files": [],
            "sync_sources": [],
            "rendering_mode": "messages",
            "model": model,
        }
        if system:
            payload["personalized_styles"] = [{"key": "custom", "instructions": system}]

        url = f"{_BASE}/api/organizations/{org}/chat_conversations/{conv_id}/completion"
        try:
            async with client.stream("POST", url, headers=_headers(session), json=payload) as resp:
                if resp.status_code >= 400:
                    body = await resp.aread()
                    raise RuntimeError(f"claude error {resp.status_code}: {body[:400]!r}")
                async for line in resp.aiter_lines():
                    if not line or not line.startswith("data:"):
                        continue
                    data = line[5:].strip()
                    if not data or data == "[DONE]":
                        continue
                    try:
                        evt = json.loads(data)
                    except json.JSONDecodeError:
                        continue
                    etype = evt.get("type")
                    if etype == "completion":
                        chunk = evt.get("completion", "")
                        if chunk:
                            yield {"delta": chunk}
                    elif etype == "content_block_delta":
                        d = evt.get("delta") or {}
                        if d.get("type") == "text_delta":
                            yield {"delta": d.get("text", "")}
                    elif etype == "message_stop":
                        break
        finally:
            try:
                await client.delete(
                    f"{_BASE}/api/organizations/{org}/chat_conversations/{conv_id}",
                    headers=_headers(session),
                )
            except Exception:
                pass


def openai_chunk(model: str, delta_text: str, finish: str | None = None) -> dict:
    return {
        "id": f"chatcmpl-{uuid.uuid4().hex[:24]}",
        "object": "chat.completion.chunk",
        "created": int(time.time()),
        "model": model,
        "choices": [{
            "index": 0,
            "delta": {"content": delta_text} if delta_text else {},
            "finish_reason": finish,
        }],
    }

File: hosted/workers/src/providers/claude.ts
import { cookieHeader, flattenContent, type ChatRequest, type SessionBlob } from "./index";

export const NAME = "claude";
export const PROVIDER_ID = "claude";
export const MODELS = [
  "claude-opus-4-5",
  "claude-sonnet-4-5",
  "claude-haiku-4-5",
  "claude-3-5-sonnet-20241022",
  "claude-3-5-haiku-20241022",
];

const BASE = "https://claude.ai";
const DOMAINS = ["claude.ai"];

function headers(session: SessionBlob): HeadersInit {
  return {
    "User-Agent": session.user_agent || "Mozilla/5.0",
    Accept: "text/event-stream, application/json",
    "Accept-Language": "en-US,en;q=0.9",
    "Content-Type": "application/json",
    Origin: BASE,
    Referer: `${BASE}/chats`,
    Cookie: cookieHeader(session, DOMAINS),
  };
}

async function organizationId(session: SessionBlob): Promise<string> {
  const r = await fetch(`${BASE}/api/organizations`, { headers: headers(session) });
  if (!r.ok) throw new Error(`claude /organizations ${r.status}`);
  const orgs = (await r.json()) as Array<{ uuid: string; capabilities?: string[] }>;
  if (!orgs.length) throw new Error("no claude organizations");
  const withChat = orgs.find((o) => (o.capabilities ?? []).includes("chat"));
  return (withChat ?? orgs[0]!).uuid;
}

function buildPrompt(messages: ChatRequest["messages"]): { system: string; prompt: string } {
  const sysParts: string[] = [];
  const convo: string[] = [];
  for (const m of messages) {
    if (m.role === "system") {
      sysParts.push(flattenContent(m.content));
      continue;
    }
    const tag = m.role === "user" ? "Human" : m.role === "assistant" ? "Assistant" : m.role;
    convo.push(`${tag}: ${flattenContent(m.content)}`);
  }
  let prompt = convo.join("\n\n");
  if (!prompt.endsWith("Assistant:")) prompt += "\n\nAssistant:";
  return { system: sysParts.join("\n\n"), prompt };
}

export async function* stream(req: ChatRequest, session: SessionBlob): AsyncIterable<{ delta: string }> {
  const model = req.model;
  const { system, prompt } = buildPrompt(req.messages);
  const org = await organizationId(session);

  const conv = crypto.randomUUID();
  const createRes = await fetch(`${BASE}/api/organizations/${org}/chat_conversations`, {
    method: "POST",
    headers: headers(session),
    body: JSON.stringify({ uuid: conv, name: "" }),
  });
  if (!createRes.ok) throw new Error(`claude conversation create ${createRes.status}`);

  const payload: Record<string, unknown> = {
    prompt,
    parent_message_uuid: "00000000-0000-4000-8000-000000000000",
    timezone: "America/Los_Angeles",
    attachments: [],
    files: [],
    sync_sources: [],
    rendering_mode: "messages",
    model,
  };
  if (system) payload.personalized_styles = [{ key: "custom", instructions: system }];

  try {
    const resp = await fetch(`${BASE}/api/organizations/${org}/chat_conversations/${conv}/completion`, {
      method: "POST",
      headers: headers(session),
      body: JSON.stringify(payload),
    });
    if (!resp.ok || !resp.body) {
      const body = await resp.text();
      throw new Error(`claude completion ${resp.status}: ${body.slice(0, 400)}`);
    }
    for await (const evt of sseEvents(resp.body)) {
      const t = evt.type;
      if (t === "completion" && typeof evt.completion === "string" && evt.completion) {
        yield { delta: evt.completion };
      } else if (t === "content_block_delta") {
        const d = (evt as { delta?: { type?: string; text?: string } }).delta;
        if (d && d.type === "text_delta" && d.text) yield { delta: d.text };
      } else if (t === "message_stop") {
        break;
      }
    }
  } finally {
    fetch(`${BASE}/api/organizations/${org}/chat_conversations/${conv}`, {
      method: "DELETE",
      headers: headers(session),
    }).catch(() => {});
  }
}

async function* sseEvents(body: ReadableStream<Uint8Array>): AsyncIterable<Record<string, unknown>> {
  const reader = body.getReader();
  const decoder = new TextDecoder();
  let buf = "";
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    buf += decoder.decode(value, { stream: true });
    let nl: number;
    while ((nl = buf.indexOf("\n")) >= 0) {
      const line = buf.slice(0, nl).trim();
      buf = buf.slice(nl + 1);
      if (!line || !line.startsWith("data:")) continue;
      const data = line.slice(5).trim();
      if (!data || data === "[DONE]") continue;
      try {
        yield JSON.parse(data);
      } catch {
        // ignore
      }
    }
  }
}
